File: Design-Patterns-Using-Go/Proxy/ProxyPatternLazy.go
package main

import "fmt"

// BookParser is an interface representing a book parser.
type BookParser interface {
	NumPages() int
}

// ConcreteBookParser is a concrete implementation of the BookParser interface.
type ConcreteBookParser struct {
	book       string
	numOfPages int
}

func NewConcreteBookParser(book string) *ConcreteBookParser {
	fmt.Println("Concrete Subject Request Method")
	// Number of pages calculation heavy operation.
	// Suppose this calculation comes to 1000 pages.
	return &ConcreteBookParser{
		book:       book,
		numOfPages: 1000,
	}
}

func (cbp *ConcreteBookParser) NumPages() int {
	fmt.Println("Concrete Subject Request Method")
	return cbp.numOfPages
}

// LazyBookParserProxy is a proxy implementation of the BookParser interface.
type LazyBookParserProxy struct {
	book    string
	subject BookParser
}

func (lbp *LazyBookParserProxy) NumPages() int {
	if lbp.subject == nil {
		lbp.subject = NewConcreteBookParser(lbp.book)
	}
	return lbp.subject.NumPages()
}

// Client code
func main() {
	proxy := &LazyBookParserProxy{book: "LOTR"}
	fmt.Println(proxy.NumPages())
}

/*
Concrete Subject Request Method
Concrete Subject Request Method
1000
*/

File: Design-Patterns-Using-Cpp/Proxy/ProxyPatternLazy.cpp
#include <iostream>

class BookParser {
public:
    virtual int numPages() = 0;
};

class ConcreteBookParser : public BookParser {
private:
    int _numPages;

public:
    ConcreteBookParser() {
        std::cout << "Concrete Book Parser Created" << std::endl;
        // Number of pages calculation heavy operation.
        // Suppose this calculation results in 1000 pages.
        this->_numPages = 1000;
    }

    int numPages() override {
        std::cout << "Concrete Book Parser Request Method" << std::endl;
        return this->_numPages;
    }
};

class LazyBookParserProxy : public BookParser {
private:
    ConcreteBookParser* subject;

public:
    LazyBookParserProxy() : subject(nullptr) {}

    int numPages() override {
        if (subject == nullptr) {
            subject = new ConcreteBookParser();
        }
        return subject->numPages();
    }

    ~LazyBookParserProxy() {
        delete subject; // Clean up allocated memory
    }
};

// Client code
int main() {
    LazyBookParserProxy proxy;
    std::cout << proxy.numPages() << std::endl;

    return 0;
}

/*
Concrete Book Parser Created
Concrete Book Parser Request Method
1000
*/
